Build guide: DNA properties (#522)
holochain/docs-pages: src/pages/build/dnas.md

* `name`: A string for humans to read. This might get used in the admin panel of Holochain [conductors](/concepts/2_application_architecture/#conductor) like [Holochain Launcher](https://github.com/holochain/launcher).
* `integrity`: Contains all the integrity modifiers for the DNA, the things that **change the DNA hash**.
    * `network_seed`: A string that serves only to change the DNA hash without affecting behavior. It acts like a network-wide passcode. {#network-seed}

>     * `properties`: Arbitrary, application-specific constants. The zome code can [read this at runtime](#use-dna-properties). Think of it as [configuration or environment variables](https://12factor.net/config) for your zomes.

Reviewer: - I think the previous wording is more accurate: they are configuration for the DNA not for zomes
- I don't think env variables is a great comparison because you can modify env variables for an existing app, whereas modifying DNA properties when cloning creates a new network.

| Alice `main`      | `call`        | `call`         | `call_remote` | ⛔            |
| Alice `search`    | `call`        | `call`         | ⛔            | `call_remote` |

## Use DNA properties

> The `properties` field in your DNA is just arbitrary bytes, but it's meant to be written and deserialized as YAML. Any of your zomes can access it, which is why it's considered modifier --- it can change the way validation functions<!--TODO: link when validation PR is merged --> operate.

Reviewer: ```suggestion
The `properties` field in your DNA is just arbitrary bytes, but it's meant to be written and deserialized as YAML. Any of your zomes can access it. Because it is a DNA modifier, it alters the DNA hash which defines the network uniqueness, and thus `properties` can only be altered before a cell is cloned or an app is installed. However, this also allows you to use `properties` in validation functions, configuring how they operate for different cells. <!-- TODO: link when validation PR is merged -->
```

I don't think my wording is great but that's what I want to get across.

You can deserialize your DNA modifiers automatically by using the [`dna_properties` macro](https://docs.rs/hdk_derive/latest/hdk_derive/attr.dna_properties.html) on a type definition, which will give your type a method called `try_from_dna_properties`.

> This example shows a helper function that only permits one agent to write data. (We call the pattern of giving one agent special privileges via DNA properties the 'progenitor' pattern.) This function could be used in a validation callback to enforce this restriction.

Reviewer: ```suggestion
This example shows a helper function that only permits one agent to write data. This function could be used in a validation callback to enforce this restriction.
```

```rust
use hdi::prelude::*;

#[dna_properties]
struct DnaProperties {

>     progenitor: AgentPubKeyB64,

Reviewer: ```suggestion
    writer: AgentPubKeyB64,
```

Author: heh

}

> fn is_allowed_to_write_data(author: AgentPubKey) -> ExternResult<bool> {

Reviewer: ```suggestion
fn is_allowed_to_write_data(author: AgentPubKeyB64) -> ExternResult<bool> {
```

Author: I suppose we could... or make them both `AgentPubKey`? My concern here was to establish patterns of ergonomic usage (unpleasant to type a Yaml-serialised `Vec<u8>`, also awkward to have to convert an `AgentPubKey` to base64 whenever you want to use this helper).

I know it's not the place to explain all that, and perhaps even doing it at all is noisy like you say. Thoughts?

Reviewer: yeah either both AgentPubKey or both AgentPubKeyB64 is fine.

>     Ok(author == DnaProperties::try_from_dna_properties()?.progenitor.try_into()?)

Reviewer: ```suggestion
    let dna_props = DnaProperties::try_from_dna_properties()?;
    Ok(author == dna_props.progenitor)
```
